graylog2-server/src/main/java/org/graylog/security/certutil/CaService.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog.security.certutil;
+
+import org.apache.commons.net.util.Base64;
+import org.glassfish.jersey.media.multipart.BodyPart;
+import org.glassfish.jersey.media.multipart.FormDataBodyPart;
+import org.graylog.security.certutil.ca.CACreator;
+import org.graylog.security.certutil.ca.exceptions.CACreationException;
+import org.graylog.security.certutil.ca.exceptions.KeyStoreStorageException;
+import org.graylog.security.certutil.keystore.storage.SmartKeystoreStorage;
+import org.graylog.security.certutil.keystore.storage.location.KeystoreFileLocation;
+import org.graylog.security.certutil.keystore.storage.location.KeystoreMongoCollections;
+import org.graylog.security.certutil.keystore.storage.location.KeystoreMongoLocation;
+import org.graylog2.Configuration;
+import org.graylog2.bootstrap.preflight.web.resources.model.CA;
+import org.graylog2.bootstrap.preflight.web.resources.model.CAType;
+import org.graylog2.plugin.system.NodeId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.CertificateException;
+import java.time.Duration;
+import java.util.List;
+import java.util.Optional;
+
+import static org.graylog.security.certutil.CertConstants.PKCS12;
+
+
+@Singleton
+public class CaService {
+    private static final Logger LOG = LoggerFactory.getLogger(CaService.class);
+    public static int DEFAULT_VALIDITY = 10 * 365;
+    private final SmartKeystoreStorage keystoreStorage;
+    private final KeystoreMongoLocation keystoreMongoLocation;
+    private final KeystoreFileLocation keystoreFileLocation;
```

Comment:
It does not have to be done now, but I'd vote for naming more explicitly what kind of keystores' locations are represented by those variables (a rash idea - `manuallyProvidedCALocation` instead of `keystoreFileLocation`)

Reply:
done